File: lib/runner.ts
// Sequential test runner: for each (endpoint, case) pair, optionally fires
// a warmup request (to absorb cold-model load time) then N timed repeats,
// back-to-back. Yields RequestMetrics as they complete so the API route
// can stream progress to the GUI live rather than waiting for the whole
// batch (important on Vercel, where a function has a hard time budget).

import { getAdapter } from "@/lib/adapters/base";
import { computeDecodeRate } from "@/lib/metrics";
import type { Endpoint, PromptCase, RequestMetrics, RunRequest } from "@/lib/types";

async function executeOnce(
  endpoint: Endpoint,
  promptCase: PromptCase,
  runIndex: number,
  isWarmup: boolean
): Promise<RequestMetrics> {
  const adapter = getAdapter(endpoint.adapter);

  const tStart = performance.now();
  let ttftMs: number | null = null;
  const textParts: string[] = [];
  let promptTokens: number | null = null;
  let completionTokens: number | null = null;
  let serverLoadS: number | null = null;
  let serverPromptEvalS: number | null = null;
  let serverEvalS: number | null = null;
  let ok = true;
  let errorMessage = "";

  try {
    for await (const event of adapter.streamChat(endpoint, promptCase)) {
      if (event.kind === "delta") {
        if (ttftMs === null) ttftMs = performance.now() - tStart;
        textParts.push(event.text);
      } else if (event.kind === "done") {
        promptTokens = event.prompt_tokens;
        completionTokens = event.completion_tokens;
        serverLoadS = event.server_load_s;
        serverPromptEvalS = event.server_prompt_eval_s;
        serverEvalS = event.server_eval_s;
      } else if (event.kind === "error") {
        ok = false;
        errorMessage = event.error_message;
      }
    }
  } catch (exc) {
    // Surface any adapter bug as a failed result, not a crashed run.
    ok = false;
    errorMessage = `${exc instanceof Error ? exc.constructor.name : "Error"}: ${exc}`;
  }

  const totalMs = performance.now() - tStart;
  const ttftS = ttftMs !== null ? ttftMs / 1000 : null;
  const totalS = totalMs / 1000;

  let serverTotalS: number | null = null;
  let proxyOverheadS: number | null = null;
  if (serverLoadS != null && serverPromptEvalS != null && serverEvalS != null) {
    serverTotalS = serverLoadS + serverPromptEvalS + serverEvalS;
    proxyOverheadS = totalS - serverTotalS;
  }

  return {
    endpoint_id: endpoint.id ?? 0,
    case_id: promptCase.id ?? 0,
    run_index: runIndex,
    is_warmup: isWarmup,
    ok,
    error_message: errorMessage,
    ttft_s: ttftS,
    total_s: totalS,
    prompt_tokens: promptTokens,
    completion_tokens: completionTokens,
    decode_tokens_per_s: computeDecodeRate(completionTokens, ttftS, totalS),
    server_load_s: serverLoadS,
    server_prompt_eval_s: serverPromptEvalS,
    server_eval_s: serverEvalS,
    server_total_s: serverTotalS,
    proxy_overhead_s: proxyOverheadS,
    response_text: textParts.join(""),
    rating: null,
    rating_notes: "",
  };
}

const sleep = (s: number) => new Promise((resolve) => setTimeout(resolve, s * 1000));

/** Runs every (endpoint, case) combination in the request, sequentially,
 * yielding each RequestMetrics as soon as it's available.
 */
export async function* runSequential(
  request: RunRequest,
  endpoints: Map<number, Endpoint>,
  cases: Map<number, PromptCase>
): AsyncGenerator<RequestMetrics> {
  for (const endpointId of request.endpoint_ids) {
    const endpoint = endpoints.get(endpointId);
    if (!endpoint) continue;

    for (const caseId of request.case_ids) {
      const promptCase = cases.get(caseId);
      if (!promptCase) continue;

      if (request.warmup) {
        // Yield the warmup result too, even though summarize() filters it
        // out of the stats. A cold model can take a minute to load, and
        // swallowing this entirely leaves the UI silent for that whole
        // time with no sign the run is alive.
        yield await executeOnce(endpoint, promptCase, -1, true);
        if (request.delay_between_s) await sleep(request.delay_between_s);
      }

      for (let i = 0; i < request.repeats; i++) {
        const metrics = await executeOnce(endpoint, promptCase, i, false);
        yield metrics;
        if (request.delay_between_s && i < request.repeats - 1) await sleep(request.delay_between_s);
      }
    }
  }
}

File: app/runner.py
"""Sequential test runner: for each (endpoint, case) pair, optionally fires
a warmup request (to absorb cold-model load time) then N timed repeats,
back-to-back. Yields RequestMetrics as they complete so the caller (the SSE
API route) can stream progress to the GUI live rather than waiting for the
whole batch.
"""
from __future__ import annotations

import asyncio
import time
from typing import AsyncIterator

import httpx

from app.adapters.base import get_adapter
from app.metrics import compute_decode_rate
from app.models import Endpoint, PromptCase, RequestMetrics, RunRequest

REQUEST_TIMEOUT_S = 300.0


async def _execute_once(
    client: httpx.AsyncClient,
    endpoint: Endpoint,
    case: PromptCase,
    run_index: int,
    is_warmup: bool,
) -> RequestMetrics:
    adapter = get_adapter(endpoint.adapter)

    t_start = time.perf_counter()
    ttft: float | None = None
    text_parts: list[str] = []
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    server_load_s: float | None = None
    server_prompt_eval_s: float | None = None
    server_eval_s: float | None = None
    ok = True
    error_message = ""

    try:
        async for event in adapter.stream_chat(client, endpoint, case):
            if event.kind == "delta":
                if ttft is None:
                    ttft = time.perf_counter() - t_start
                text_parts.append(event.text)
            elif event.kind == "done":
                prompt_tokens = event.prompt_tokens
                completion_tokens = event.completion_tokens
                server_load_s = event.server_load_s
                server_prompt_eval_s = event.server_prompt_eval_s
                server_eval_s = event.server_eval_s
            elif event.kind == "error":
                ok = False
                error_message = event.error_message
    except Exception as exc:  # noqa: BLE001 - surface any adapter bug as a failed result, not a crashed run
        ok = False
        error_message = f"{type(exc).__name__}: {exc}"

    total = time.perf_counter() - t_start

    server_total_s = None
    proxy_overhead_s = None
    if server_load_s is not None and server_prompt_eval_s is not None and server_eval_s is not None:
        server_total_s = server_load_s + server_prompt_eval_s + server_eval_s
        proxy_overhead_s = total - server_total_s

    return RequestMetrics(
        endpoint_id=endpoint.id or 0,
        case_id=case.id or 0,
        run_index=run_index,
        is_warmup=is_warmup,
        ok=ok,
        error_message=error_message,
        ttft_s=ttft,
        total_s=total,
        prompt_tokens=prompt_tokens,
        completion_tokens=completion_tokens,
        decode_tokens_per_s=compute_decode_rate(completion_tokens, ttft, total),
        server_load_s=server_load_s,
        server_prompt_eval_s=server_prompt_eval_s,
        server_eval_s=server_eval_s,
        server_total_s=server_total_s,
        proxy_overhead_s=proxy_overhead_s,
        response_text="".join(text_parts),
    )


async def run_sequential(
    request: RunRequest,
    endpoints: dict[int, Endpoint],
    cases: dict[int, PromptCase],
) -> AsyncIterator[RequestMetrics]:
    """Runs every (endpoint, case) combination in request, sequentially,
    yielding each RequestMetrics as soon as it's available.
    """
    async with httpx.AsyncClient(timeout=REQUEST_TIMEOUT_S) as client:
        for endpoint_id in request.endpoint_ids:
            endpoint = endpoints[endpoint_id]
            for case_id in request.case_ids:
                case = cases[case_id]

                if request.warmup:
                    await _execute_once(client, endpoint, case, run_index=-1, is_warmup=True)
                    if request.delay_between_s:
                        await asyncio.sleep(request.delay_between_s)

                for i in range(request.repeats):
                    metrics = await _execute_once(client, endpoint, case, run_index=i, is_warmup=False)
                    yield metrics
                    if request.delay_between_s and i < request.repeats - 1:
                        await asyncio.sleep(request.delay_between_s)
